src/detect.rs

```diff
@@ -18,3 +23,196 @@ pub enum DetectOutcome {
     Pass(BuildPlan),
     Fail,
 }
+
+enum BuildpackTomlArg<BM> {
+    Path(PathBuf),
+    Str(String),
+    BuildpackToml(BuildpackToml<BM>),
+}
+
+enum PlatformArg<P> {
+    Plat(P),
+    Dir(PathBuf),
+}
+
+pub struct DetectContextBuilder<P: Platform, BM: DeserializeOwned> {
+    app_dir: Option<PathBuf>,
+    buildpack_dir: Option<PathBuf>,
+    stack_id: String,
+    platform: Option<PlatformArg<P>>,
+    buildpack_toml: Option<BuildpackTomlArg<BM>>,
+}
+impl<P: Platform, BM: DeserializeOwned> DetectContextBuilder<P, BM> {
+    pub fn new(stack_id: impl AsRef<str>) -> Self {
+        DetectContextBuilder {
+            app_dir: None,
+            buildpack_dir: None,
+            stack_id: String::from(stack_id.as_ref()),
+            platform: None,
+            buildpack_toml: None,
+        }
+    }
+
+    pub fn buildpack_toml(mut self, arg: BuildpackToml<BM>) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::BuildpackToml(arg));
+        self
+    }
+
+    pub fn buildpack_toml_from_string(
+        mut self,
+        string: impl AsRef<str>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::Str(string.as_ref().to_string()));
+        self
+    }
+
+    pub fn buildpack_toml_from_path(
+        mut self,
+        path: impl AsRef<Path>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::Path(path.as_ref().to_path_buf()));
+        self
+    }
+
+    /// Used for testing purposes to set a minimal viable context
+    ///
+    /// Warning: The directory passed in will be modified.
+    ///
+    ///```
+    /// use tempfile;
+    /// use libcnb::DetectContextBuilder;
+    ///
+    /// pub fn detect(context: libcnb::GenericDetectContext) {}
+    ///
+    /// let temp_dir = tempfile::tempdir().unwrap();
+    /// let context = DetectContextBuilder::new("heroku-20")
+    ///     .buildpack_toml_from_string(include_str!("../examples/example-02-ruby-sample/buildpack.toml"))
+    ///     .from_path(&temp_dir)
+    ///     .build::<std::io::Error>()
+    ///     .unwrap();
+    ///
+    /// detect(context);
+    ///```
+    pub fn from_path(mut self, dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
```

Comment:
In Rust, `from_` is[ a widely used prefix](https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case) for converting `A` into `B`. This function does something very different and might be confusing since it goes against that convention.

Reply:
I feel like we don't need this function at all. It could be default behaviour to create temporary directories for `app_dir` and `platform_dir`. It would also simplify the amount of state we need to keep track of in the builder itself.

In addition, the way this is implemented currently, a builder cannot be safely used concurrently since, when using temporary directories created by this function, every invocation of build will create contexts with the same paths.

Reply:
>  It could be default behavior to create temporary directories for app_dir and platform_dir. 

This requirement of the builder to create the temporary does two things, it forces us to have `tempfile` as a prod dependency instead of runtime. (I'm fine with it, but its a consequence). It also would force us into the pattern I mentioned before where a builder must be stored as a variable before we can generate a context from it and it won't be able to be used as a one-liner.

That's not so different from my suggested TempContext API:

```
let temp = TempContext::new(
    "heroku-20", 
    include_str!("../../examples/example-02-ruby-sample/buildpack.toml")
); // <== Must assign to a variable to prevent tempdir from being cleaned
let build_context = temp.build;
```

Ultimately I'm fine either way. Both methods (dependency injection of temp dir versus generating a temp dir) require at least one assignment. So from the end-user experience, it's about the same.

Maybe there's some way to annotate that I expect the lifetime of the temporary directory to outlast the returned Context object? I'm unsure.

> a builder cannot be safely used concurrently since, when using temporary directories created by this function, every invocation of build will create contexts with the same paths.

Right now there's a 1-1 mapping of builders and contexts. The builder cannot be re-used after it generates a context as it is a consuming builder. I am in favor of making P and BM cloneable which would enable us to to generate multiple contexts from the same builder.  But I've not tried it.

> a builder cannot be safely used concurrently since, when using temporary directories created by this function, every invocation of build will create contexts with the same paths.

As I was writing this I was imagining that if you don't want anything shared with another context you would need to generate another builder. Right now that's a hard requirement due to the consuming nature of the builder. But if that changes it could change the way people want to use it.

I was thinking one builder per test. I wasn't thinking of trying to set some sort of global static value that could be called again and again. 

With that in mind using an enum like `Temporary` and generating the directory on demand in the `build()` call would allow for builder re-use which it seems like is a goal of yours. I can try out that tactic, but it won't pay off unless we can also get clonable `P` and `BM`.

> In Rust, from_ is a widely used prefix for converting A into B. This function does something very different and might be confusing since it goes against that convention.

I agree in terms of the naming being misleading. If I switch back to generating temporary directories internally and using an enum then It could be something like `with_tempdir()` and take no arguments.

Reply:
I've updated so the builder can generate contexts via oneliners and can be re-used. It now also generates and retains temp dirs.